io/genbank/parser.go

```diff
@@ -0,0 +1,356 @@
+	if err != nil {
+		return res, p.makeSyntaxError("failed to parse date", err)
+	}
+	res.Date = date
+
+	p.readLine() // Skip empty line
+
+	releaseLine, err := p.readLine()
+	if err != nil {
+		return res, fmt.Errorf("failed to read release: %w", err)
+	}
+	if len(releaseLine) < 48 {
+		return res, p.makeSyntaxError("release line too short")
+	}
+	release := strings.Split(strings.TrimSpace(releaseLine[47:]), ".")
+	if len(release) != 2 {
+		return res, p.makeSyntaxError("malformed release version")
+	}
+	res.MajorRelease, err = strconv.Atoi(release[0])
+	if err != nil {
+		return res, p.makeSyntaxError("failed to parse major release", err)
+	}
+	res.MinorRelease, err = strconv.Atoi(release[1])
+	if err != nil {
+		return res, p.makeSyntaxError("failed to parse minor release", err)
+	}
+
+	p.readLine() // Skip empty line
+
+	titleLine, err := p.readLine()
+	if err != nil {
+		return res, fmt.Errorf("failed to read title: %w", err)
+	}
+	res.Title = strings.TrimSpace(titleLine)
+
+	p.readLine() // Skip empty line
+
+	statsLine, err := p.readLine()
+	if err != nil {
+		return res, fmt.Errorf("failed to read file statistics: %w", err)
+	}
+	if len(releaseLine) < 48 {
+		return res, p.makeSyntaxError("release line too short")
+	}
+	res.NumEntries, err = strconv.Atoi(strings.TrimSpace(statsLine[:8]))
+	if err != nil {
+		return res, p.makeSyntaxError("failed to parse number of entries/loci", err)
+	}
+	res.NumBases, err = strconv.Atoi(strings.TrimSpace(statsLine[15:26]))
+	if err != nil {
+		return res, p.makeSyntaxError("failed to parse number of bases", err)
+	}
+	res.NumSequences, err = strconv.Atoi(strings.TrimSpace(statsLine[39:47]))
+	if err != nil {
+		return res, p.makeSyntaxError("failed to parse number of sequences", err)
+	}
+
+	p.readLine() // Skip empty line
+
+	return res, nil
+}
+
+func (p *Parser) parseEntry() (entry Entry, reachedEOF bool, err error) {
+	keywordFuncs := map[string]func(context) error{
+		"LOCUS": p.parseLocus,
+	}
+
+	res := Entry{}
+	ctx := context{entry: &res}
+	for {
+		lineType, err := p.peekNextLineType()
+		if err != nil {
+			return res, false, fmt.Errorf("could not parse entry: %w", err)
+		}
+
+		switch lineType {
+		case keyword:
+			err = p.dispatchByPrefix(ctx, keywordFuncs)
+			if err == io.EOF {
+				return res, true, nil
+			} else if err != nil {
+				return res, false, err
+			} else if _, err := p.peekLine(); err == io.EOF {
+				return res, true, nil
+			}
+
+		case empty:
+			p.readLine() // We need to read the empty line to advance the reader
+		case eof:
+			return res, true, nil
+		case subKeyword:
+			return res, false, p.makeSyntaxError("subkeyword found outside of keyword context")
+		case entryEnd:
+			return res, false, nil
+		case continuation:
+			return res, false, p.makeSyntaxError("continuation found outside of keyword context")
+		case featureKey:
+			return res, false, p.makeSyntaxError("feature key found outside of FEATURES keyword context")
```

Comment:
These static errors would be common across the different contexts right? Maybe it is a bit premature, but it might be nice to have a default case with something like

```go
func GetWrongContextError(lineType LineType) error {
    switch lineType {
    // the things
    default: 
        // something went really wrong 
    }
}

That way it can be shared for the other parse functions for the other contexts.

Reply:
Great idea! Done.

Reply:
Regardless, I think it's better for individual functions to bump up to the main `parseEntry` loop if they encounter a linetype they don't recognize. That way, error handling is all in one place.

Reply:
Sounds good to me!